Crash trying to commit a tx log

After tying to proccess an L2 Bridge (Deposit in L2 to withdrawal fund to L1) with a low gas estimation returned by the hermez core
The gas estimation log is here:
```
2022-04-11T08:48:54Z    DEBUG   synchronizer/synchronizer.go:168        Getting rollup info from block 36 to block 136
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method net_version id 15 params 
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getBalance id 16 params ["0xc949254d682d8c9ad5682521675b8f43b102aec4","latest"]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method net_version id 17 params 
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getBlockByNumber id 18 params ["latest",false]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_gasPrice id 19 params 
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getCode id 20 params ["0x9d98deabc42dd696deb9e40b4f1cab7ddbf55988","pending"]

2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_estimateGas id 21 params [{"data":"0x0e21fbd700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c949254d682d8c9ad5682521675b8f43b102aec4","from":"0xc949254d682d8c9ad5682521675b8f43b102aec4","gasPrice":"0x3b9aca00","to":"0x9d98deabc42dd696deb9e40b4f1cab7ddbf55988","value":"0xde0b6b3a7640000"}]

2022-04-11T08:48:54Z    DEBUG   state/host.go:97        GetCodeHash for address 0x9D98DeAbC42dd696Deb9e40b4f1CAB7dDBF55988 => 0x79fe968d588abb9296dc26ab425184bbe8f2074fc1ee2f3e4e5c174bba0687e0
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getTransactionCount id 22 params ["0xc949254d682d8c9ad5682521675b8f43b102aec4","pending"]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_sendRawTransaction id 23 params ["0xf8f202843b9aca00825b98949d98deabc42dd696deb9e40b4f1cab7ddbf55988880de0b6b3a7640000b8840e21fbd700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c949254d682d8c9ad5682521675b8f43b102aec48207f6a0125b225d831a8ce9ee706ffd1fe89ec3840112a0a648f7dc11edf4a9b33ddfd0a06e9a843753394dcc9b7497b7723ffb7417998470b2f765b428659fd7622836c8"]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/eth.go:376      adding TX to the pool: 0xce6f296c94b656d0cf08f062eea7cd3adb46f433f0f1010614a113b7d0095cb2
2022-04-11T08:48:54Z    DEBUG   jsonrpc/eth.go:381      TX added to the pool: 0xce6f296c94b656d0cf08f062eea7cd3adb46f433f0f1010614a113b7d0095cb2
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getBlockByNumber id 140 params ["0x2",true]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getBlockByNumber id 141 params ["latest",false]
2022-04-11T08:48:54Z    DEBUG   jsonrpc/handler.go:55   request method eth_getLogs id 142 params [{"address":["0x9d98deabc42dd696deb9e40b4f1cab7ddbf55988"],"fromBlock":"0x3","toBlock":"0x67","topics":null}]
2022-04-11T08:48:55Z    DEBUG   synchronizer/synchronizer.go:168        Getting rollup info from block 36 to block 136
```

Here is the log of the crash:
```
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        DUP3
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        PUSH1
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        DUP2
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        LT
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        PUSH2
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        JUMPI
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        JUMPDEST
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        ADD
2022-04-11T08:48:58Z    DEBUG   evm/state.go:244        SLOAD
2022-04-11T08:48:58Z    DEBUG   state/batchprocessor.go:248     Transaction Data [14 33 251 215 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 13 224 182 179 167 100 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 201 73 37 77 104 45 140 154 213 104 37 33 103 91 143 67 177 2 174 196]
2022-04-11T08:48:58Z    DEBUG   state/batchprocessor.go:249     Returned value from execution: 0x
2022-04-11T08:48:58Z    WARN    state/batchprocessor.go:114     Error processing transaction 0xce6f296c94b656d0cf08f062eea7cd3adb46f433f0f1010614a113b7d0095cb2: out of gas
2022-04-11T08:48:58Z    ERROR   state/batchprocessor.go:115     gas: 23448 gasPrice: 1000000000
2022-04-11T08:48:58Z    ERROR   state/batchprocessor.go:608     commit:  {Address:0x9D98DeAbC42dd696Deb9e40b4f1CAB7dDBF55988 Topics:[0xb9cc79b4411bd33ec9620f8fe3d6236a5d18050ca48e991864dc0556a83d1a2b] Data:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 13 224 182 179 167 100 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 201 73 37 77 104 45 140 154 213 104 37 33 103 91 143 67 177 2 174 196 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] BlockNumber:3 TxHash:0xce6f296c94b656d0cf08f062eea7cd3adb46f433f0f1010614a113b7d0095cb2 TxIndex:0 BlockHash:0xddc770f982827d35d1349be3dcfa54d5e8654e6c249f83f8f8608542d09a476c Index:0 Removed:false}
2022-04-11T08:48:58Z    FATAL   synchronizer/synchronizer.go:276        error processing batch. BatchNumber: 3. Error: ERROR: insert or update on table "log" violates foreign key constraint "log_transaction_hash_fkey" (SQLSTATE 23503)
/src/log/log.go:106 github.com/hermeznetwork/hermez-core/log.appendStackTraceMaybeArgs()
/src/log/log.go:137 github.com/hermeznetwork/hermez-core/log.Fatal()
/src/synchronizer/synchronizer.go:276 github.com/hermeznetwork/hermez-core/synchronizer.(*ClientSynchronizer).processBlockRange()
/src/synchronizer/synchronizer.go:178 github.com/hermeznetwork/hermez-core/synchronizer.(*ClientSynchronizer).syncBlocks()
/src/synchronizer/synchronizer.go:112 github.com/hermeznetwork/hermez-core/synchronizer.(*ClientSynchronizer).Sync.func1()
```

--- a/state/batchprocessor.go
+++ b/state/batchprocessor.go
@@ -79,7 +79,7 @@ func (b *BasicBatchProcessor) ProcessBatch(ctx context.Context, batch *Batch) er
 	var index uint
 
 	b.CumulativeGasUsed = 0
-	b.Host.logs = []types.Log{}
+	b.Host.logs = map[common.Hash][]*types.Log{}
 
 	// Set Global Exit Root storage position
 	var batchNumberBuf, storagePositionBuf [32]byte
@@ -263,12 +263,7 @@ func (b *BasicBatchProcessor) populateBatchHeader(batch *Batch) {
 	rr := make([]*types.Receipt, 0, len(batch.Receipts))
 	for _, receipt := range batch.Receipts {
 		r := receipt.Receipt
-		for _, l := range b.Host.logs {
-			if l.TxHash == receipt.TxHash {
-				rl := l
-				r.Logs = append(r.Logs, &rl)
-			}
-		}
+		r.Logs = append(r.Logs, b.Host.logs[receipt.TxHash]...)
 		rr = append(rr, &r)
 	}
 	block := types.NewBlock(batch.Header, batch.Transactions, batch.Uncles, rr, &trie.StackTrie{})
@@ -566,12 +561,16 @@ func (b *BasicBatchProcessor) commit(ctx context.Context, batch *Batch) error {
 	}
 
 	// store logs
-	for _, txLog := range b.Host.logs {
-		txLog.BlockHash = blockHash
-		txLog.BlockNumber = batch.Number().Uint64()
-		err := b.Host.State.AddLog(ctx, txLog)
-		if err != nil {
-			return err
+	for _, tx := range batch.Transactions {
+		if txLogs, found := b.Host.logs[tx.Hash()]; found {
+			for _, txLog := range txLogs {
+				txLog.BlockHash = blockHash
+				txLog.BlockNumber = batch.Number().Uint64()
+				err := b.Host.State.AddLog(ctx, *txLog)
+				if err != nil {
+					return err
+				}
+			}
 		}
 	}
 

--- a/state/host.go
+++ b/state/host.go
@@ -17,7 +17,7 @@ type Host struct {
 	State              *State
 	stateRoot          []byte
 	transactionContext transactionContext
-	logs               []types.Log
+	logs               map[common.Hash][]*types.Log
 	forks              runtime.ForksInTime
 	runtimes           []runtime.Runtime
 }
@@ -167,18 +167,31 @@ func (h *Host) GetBlockHash(number int64) common.Hash {
 func (h *Host) EmitLog(address common.Address, topics []common.Hash, data []byte) {
 	if !h.transactionContext.simulationMode {
 		log.Debugf("EmitLog for address %v", address)
-		txLog := types.Log{
+
+		txLog := &types.Log{
 			Address: address,
 			Topics:  topics,
 			Data:    common.CopyBytes(data),
 			TxHash:  h.transactionContext.currentTransaction.Hash(),
 			TxIndex: h.transactionContext.index,
-			Index:   uint(len(h.logs)),
+			Index:   h.getLogIndex(),
 			Removed: false,
 		}
 
-		h.logs = append(h.logs, txLog)
+		if _, found := h.logs[h.transactionContext.currentTransaction.Hash()]; !found {
+			h.logs[h.transactionContext.currentTransaction.Hash()] = []*types.Log{}
+		}
+
+		h.logs[h.transactionContext.currentTransaction.Hash()] = append(h.logs[h.transactionContext.currentTransaction.Hash()], txLog)
 	}
+}
+
+func (h *Host) getLogIndex() uint {
+	nextIndex := 0
+	for l := range h.logs {
+		nextIndex += len(l)
+	}
+	return uint(nextIndex)
 }
 
 // Callx calls a SC
@@ -354,7 +367,11 @@ func (h *Host) setRuntime(r runtime.Runtime) {
 func (h *Host) run(ctx context.Context, contract *runtime.Contract) *runtime.ExecutionResult {
 	for _, r := range h.runtimes {
 		if r.CanRun(contract, h, &h.forks) {
-			return r.Run(ctx, contract, h, &h.forks)
+			result := r.Run(ctx, contract, h, &h.forks)
+			if result.Err != nil {
+				delete(h.logs, h.transactionContext.currentTransaction.Hash())
+			}
+			return result
 		}
 	}
 

